R/install.R

```diff
@@ -1,276 +1,343 @@
 
-  if (is.null(plan)) {
-    res <- lapply(seq_along(filenames), function(idx) {
-      file <- filenames[idx]
-      meta <- metadata[[idx]]
-      format_message <- inplace(function(x) bar$message(format(x)))
-      installed_path <- with_handlers(
-        pkginstall_installed = format_message,
-        pkginstall_built = format_message,
-        pkginstall_begin = format_message,
-        error = exiting(function(e) {
-          name <- basename(e$package %||% e$path %||% file)
-          bar$message(glue("{red_cross()} Failed {name}"))
-          stop(e)
-        }), {
-          bar$tick(0, tokens = list(packages = get_pkg_name(file)))
-          install_package(file, lib = lib, metadata = meta,
-                          vignettes = vignettes)
-        })
-
-      bar$tick(1, tokens = list(packages = get_pkg_name(file)))
-      installed_path
-    })
-
-    names(res) <- filenames
-    return(structure(res, class = "installation_results", elapsed = Sys.time() - start))
-  }
+    ## Initialise one task for each worker
+    for (i in seq_len(state$config$num_workers)) {
+      ## TODO: update progress bar here
+      task <- select_next_task(state)
+      state <- start_task(state, task)
+    }
 
-  running <- character()
-  results <- list()
-  processes <- list()
-  update_progress <- function(count) {
-    bar$tick(count, tokens = list(packages = collapse(running, ", "), processes = length(processes), num_workers = num_workers))
-  }
-  update_progress(0)
-
-  # Currently assumes the order of installation is not important
-  events <- get_events(list(plan = plan, processes = list(), results = list()), num_workers, lib, bar)
-  while(running_processes(events)) {
-    for (output in get_output(events)) {
-      is_running <- grepl("^Building", strip_style(output$output))
-      running <- union(running, sub("^Building ", "", strip_style(output$output[is_running])))
-      finished <- output$output[!is_running]
-
-      if (length(finished) > 0 && nzchar(finished)) {
-
-        # remove installed from running
-        installed <- strip_style(finished)[grepl("^. Installed ", strip_style(finished))]
-        installed <- sub(". Installed ([^[:space:]]+).*", "\\1", installed)
-
-        running <- setdiff(running, installed)
-        bar$message(finished)
-        update_progress(length(installed))
-      }
-      if (!bar$finished && length(running)) {
-        update_progress(0)
-      }
+    while (1) {
+      if (are_we_done(state)) break;
+      ## TODO: update progress bar here
+
+      events <- poll_workers(state)
+      state <- handle_events(state, events)
+      task  <- select_next_task(state)
+      state <- start_task(state, task)
     }
-    events <- get_events(events, num_workers, lib, bar)
+  }, error = function(e) kill_all_processes(state))
+
+  create_install_result(state)
+}
+
+make_start_state <- function(plan, config) {
+
+  ## We store the data about build and installation here
+  install_cols <- data.frame(
```

Comment:
We have tibble as an import already, so this could be `tibble::data_frame()` with the `I()` calls removed I think.

Reply:
I removed tibble I think, but I don't mind, I think pkgdepends uses it already, anyway.

Reply:
Actually, at some point, I would like to trim the deps, because every time you are installing sg that conflicts with a dep of pkgman, pkgman breaks.